Storage reversal, alternative to #1513

aprokop opened the pull request:
@trilinos/ifpack2 @trilinos/tpetra @mhoemmen @sthomas61

Seems to work, but multiple improvements required:

- [x] See #1523 
  Could be bypassed if Tpetra provided a `CrsMatrix` constructor taking in all 4 maps and a Kokkos matrix. The only Kokkos matrix-based constructor takes in only row and column maps at this point.
- [x] Call `setMatrix` for `U_solver_` and `L_solver_` in RILUK's `compute()` so that the solver storage is updated
- [x] Remove pointer equal shortcut `LocalSparseTriangularSolver::setMatrix` so that the storage is updated (otherwise, if matrix values are updated but stored in the same matrix, these updates are ignored)
- [x] ~Need to move some stuff to parallel regions, if serial works out well~

aprokop commented:
OK, I implemented a Tpetra matrix/graph constructors taking in 4 maps and a local matrix. This could/should probably go into a separate PR, but at least it works around of the first issue.

sthomas61 commented:
Andrew - the storage reversal could still improve L1-cache hits and speed- it can be applied after the level-set algorithm(s) ?

ambrad commented:
@sthomas6, @aprokop, HTS (implicitly) reverses the order of entries in the case of U. In more detail, HTS does a fair bit of repacking; it doesn't see U as something special in itself, but rather it's one detail among many in determining how memory is packed. The end result is that in the solve phase, each thread marches linearly through memory, through each level, and then through the recursively blocked dense remainder.

aprokop commented:
@ambrad If HTS also stores some internal state, then the check in `setMatrix` should also be ignored for HTS, right? Same pointer for a matrix does not mean that the matrix hasn't changed.

ambrad commented:
@aprokop, I'm fine with that change. The setMatrix doc says initialize() *must* be called after calling setMatrix, so that pointer check I guess is pointless (but note other preconditioners do the same). Looking through RILUK, it appears that setting htsImpl_ to null always in setMatrix shouldn't affect performance; RILUK calls setMatrix in initialize(), which is fine, and from its own setMatrix, which has the pointer comparison check. So feel free to reset the HTS pointer.

mhoemmen commented:
@aprokop I will open a separate issue for the new constructors.

mhoemmen commented:
> The setMatrix doc says initialize() must be called after calling setMatrix, so that pointer check I guess is pointless (but note other preconditioners do the same).

I would prefer requiring users to call initialize() themselves.  In fact, I dislike the current behavior that calls initialize() automatically in compute() if initialize() hasn't yet been called.

mhoemmen commented:
@aprokop @sthomas61 What's the scoop?  Is this ready for merge or should we wait a bit?

aprokop commented:
I'll rebase this PR on top of #1547 once that is merged, and then this can also be merged.

aprokop commented:
OK. I rebased and cleaned up this PR.

I'm not sure whether the new code should be exposed unconditionally, as it is not tested in applications. Should I hide it behind some CMake macro? I could also try to see if I could make it through a parameter list, but this may not work, depending on how triangular solve uses (in RILUK, ILUT, and such) pass their parameter lists to it.

mhoemmen commented:
@aprokop wrote:
> I'm not sure whether the new code should be exposed unconditionally, as it is not tested in applications. Should I hide it behind some CMake macro?

I'm not keen on adding more untested code paths.  If we want this to become the default, there are two questions:

  1. Is this generally always a good idea for performance?
  2. Does it increase set-up time or memory usage significantly?

It looks like the answer to (1) is "yes, this is generally a good idea."  This idea of reversing U comes from the PETSc folks ( http://dl.acm.org/citation.cfm?id=2076558 ).  They optimized specifically for the case of ILU, which is the case of interest for us as well.  For complete factorizations, either the implementation uses its own sparse triangular solve (e.g., SuperLU), or we already have an optimized, thread-parallel implementation tuned for complete factorizations, namely HTS.  @ambrad wrote above that "HTS reverses the order of entries in the case of U" as well.  Thus, we're only doing what multiple smart people have already done :-) .

Regarding (2), Ifpack2 already stores the L and U factors separately.  The reversal of U may increase the memory high-water mark.  However, `Ifpack2::AdditiveSchwarz` overlap already uses a lot more memory than it could.  I'm not sure if we have optimized Ifpack2's ILU(k) at all.  Thus, we're not adding much more than the thing already uses.  Furthermore, the AdditiveSchwarz overlap high-water mark happens at a different time than the sparse factorization, and involves temporary storage (the intermediate overlap matrices) that does not coexist in memory with the sparse factorization results.

I would say: Add a ParameterList run-time option to turn this off, but turn it on by default.  At the very least, this thing should always _build_.  This is not some little summer research project; this is code that we intend for production use.  If we're not sure and want to do some performance testing first, we could always start with the run-time option turned off.

mhoemmen commented on packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_decl.hpp:
```diff
@@ -107,6 +107,9 @@ class LocalSparseTriangularSolver :
                             global_ordinal_type, node_type> row_matrix_type;
+  //! Specialization of Tpetra::CrsMatrix used by this class.
+  typedef Tpetra::CrsMatrix<scalar_type, local_ordinal_type,
+                            global_ordinal_type, node_type, false> crs_matrix_type;
```
You may leave off the fifth template parameter `false`.  We intend to get rid of that at some point, and it has the right default value.

aprokop replied:
OK, got rid of it.

aprokop replied:
Removing `false` does not work as compilation fails:
```
wrong number of template arguments (4, should be 5)
```

mhoemmen replied:
@aprokop Ah, it's a forward declaration issue.  Nalu has this issue too.  Never mind then :-) This reminds me that Tpetra will need forward declaration headers at some point.

mhoemmen commented on packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_def.hpp:
```diff
@@ -358,8 +359,75 @@ initialize ()
+
+    typename decltype(ptr)::non_const_type  newptr ("ptr", ptr.dimension_0 ());
+    typename decltype(ind)::non_const_type  newind ("ind", ind.dimension_0 ());
+    decltype(val)                           newval ("val", val.dimension_0 ());
```
This should work for now, because we assume UVM.  However, please do insert an `execution_space::fence()` before and after the setup phase, just to make sure that the GPU sees the changes made on host.  Also, please add a comment saying something like "FIXME this assumes UVM."  Thanks!

mhoemmen commented on packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_def.hpp:
```diff
@@ -358,8 +359,75 @@ initialize ()
+
   if (Teuchos::nonnull (htsImpl_))
+  {
     htsImpl_->initialize (*A_crs_);
```
Have we tested with the HTS option enabled?  I realized that HTS also wants to reverse U, and I'm not sure if the two reversals will fight each other.  We should make sure that the test exercises a matrix whose reversal is not itself, e.g., not the usual [-1 2 -1] matrix.

aprokop replied:
HTS option is orthogonal to this code. Specifically, the check for reversal is:
```c++
  if (A_crs_->isUpperTriangular() && htsImpl_.is_null()) {
```
So, when `htsImpl_` is enabled (and it should be done before `initialize()`) we don't do reversal.
In any case, I will do the HTS testing.

mhoemmen replied:
Ah, OK, never mind then :-)  Not sure how hard it is to set up HTS testing.  HTS lives in ShyLU for Various Reasons.  I lost that debate ;-) .  LinearSolverFactory never took off and we never figured out how to get it to work right with static libraries, so we had to put HTS _somewhere_.

ambrad replied:
@mhoemmen, if ShyLU/HTS is enabled, then Belos and units tests in Ifpack2 for HTS get enabled, too.

mhoemmen commented on packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_def.hpp:
```diff
@@ -358,8 +359,75 @@ initialize ()
+
+      Teuchos::Array<global_ordinal_type> newRowElems(rowElems.size());
+      for (size_t i = 0; i < numElems; i++)
+        newRowElems[i] = rowElems[numElems-1 - i];
```
Do we actually need the reversed-storage matrix to be a `Tpetra::CrsMatrix` with proper Maps?  The local sparse triangular solver should only need the `KokkosSparse::CrsMatrix`.  I'm asking because Map construction takes time and memory, but doesn't appear necessary for the local solves.

aprokop replied:
I find this way is a lot easier, as the permutation of input and output vectors is done automatically. I think that `apply` may also be slightly faster as instead of `gather vector through import + permute it locally` we only do `gather vector through import into an already permuted storage`. Same with export. Plus, I think the code is cleaner this way as the whole permutation is confined within the `initialize` call.

mhoemmen replied:
@aprokop That's fine.  Those Map constructors don't need communication so the only cost should be computing the reversed GIDs and setting up the hash table.

aprokop commented:
@mhoemmen Nalu results (for non-threaded runs) demonstrate dramatic increase in performance with this patch when used with MueLu with ILUT smoothing. And yes, PETSc folks have moved towards it. As the code is self-contained in Ifpack2, it will only be used for ILU-type smoothers, and does not touch any of the triangular solves of the direct solvers. It also does not affect HTS, which already does the reversal.

Regarding (2), I think you would see increase in storage for Additive Schwarz without overlap, as most of the involved components (LocalFilter, OverlappedMatrix and such) just hijack the original storage and do not introduce anything else. So if you do ILU, you would have 1x (original matrix) + 1x (L+U) factors + 1/2x (reversed U factor). Compared to 2x, this is 25% increase. For overlapped Schwarz, this is going to be smaller. Theoretically, this could be removed if one got rid of storing L and U factors inside the ILU methods like ILUT and RILUK.

> I would say: Add a ParameterList run-time option to turn this off, but turn it on by default.
I would cautious to turn it on by default. I would start with off, and advertise to selected users. If they don't hit any problems, then we can turn it on. This code is critical, so no reason to irritate users with an early code.

mhoemmen commented:
@sthomas61 reported extraordinary performance results in Nalu, to the point where it's worth considering ILU as a smoother, even given the modest extra set-up time.  I trust your discretion :-)

aprokop commented:
OK< I added parameter propagation.

mhoemmen commented:
Thanks @aprokop !  Recreating the solver in compute() could introduce some overhead for a more elaborate sparse triangular solver, but it's probably fine for now.

aprokop commented:
@mhoemmen wrote:

> Recreating the solver in compute() could introduce some overhead for a more elaborate sparse triangular solver

I'm not sure I understand. Could you please give an example?

mhoemmen commented:
@aprokop wrote:
> I'm not sure I understand. Could you please give an example?

Suppose that we use a TPL to implement sparse triangular solves.  It might have separate symbolic and numeric set-up phases.  Anyway, not a big deal :-)

aprokop commented:
@mhoemmen Oh, I agree with that. However, I don't think this is a LocalSparseTriangularSolve limiation. Rather, for instance, ILUT should follow that notion and *not* construct L and U graphs in `compute()`.

mhoemmen commented:
@aprokop Cool, no worries then :-)

sthomas61 commented:
One reason that Nalu speeds up so much for the ABL case with reversed U may be the low continuity GMRES iteration count. ~5-7 iterations with ILUT. 

whereas the McAlister blade problem has 40-50 iterations - perhaps matrix-vector multiply then dominates

aprokop commented:
@sthomas61 has kindly agreed to test this PR with Nalu to make sure I did not screw up anything during the cleanup. Once that's done, and we get similar results, I'll run the checkin script.

sthomas61 commented:
Andrey's branch validates - produces the same results that I have.
The HTS option also works and gives similar results but is slower in serial
MPI+1 compared to the storage reversed U alone

aprokop commented:
I'm running the checkin script, and will close this PR after the push.

sthomas61 commented:
ABL 256^3 benchmark on 384 cores Peregrine: reversal of U
STKPERF: Total Time: 104.5811

compared to HTS:
STKPERF: Total Time: 105.7059

sthomas61 commented:
ABL 256^3 benchmark on 96 cores Peregrine without reversal, > 500 seconds
with reversal of U:
STKPERF: Total Time: 376.8727

compared to HTS:
STKPERF: Total Time: 399.4143

aprokop commented:
OK, pushed in 541e11b00a8f0ad0d598c7a144a76e81f0a30d1f. Thanks to everybody!

ambrad commented:
@aprokop, I might be wrong, but if you're using ILUT in particular, it might be worth looking at the ILUT::apply() impl a bit, outside of just the LocalSparseTri apply() calls. I suspect there's at least one memory alloc every single apply() call, whereas IIRC we got RILUK::apply() down to none in the common case.

aprokop commented:
Thanks, @ambrad, I'll take a look.

mhoemmen commented:
Thanks @aprokop and @sthomas61 !!!  Special thanks to @sthomas61 for taking the initiative to investigate this.
